README.md
# RoMiner

This is the RoMiner server. But what is RoMiner?

RoMiner is/was a service that allowed users to mine cryptocurrency on their machines and in return they would recieve Robux(currency used on Roblox.com which is a very large game platform).

Server functions:
Accept new-user signups
Continuously update user balances based on the amount of crypto mined
Process withdrawal requests made by users


history.txt
poopants of 2 to 18257354
xxflick of 57 to 31624774
xxflick of 63 to 31624774
youtubevideo of 256 to 32871316
youtubevideo of 41 to 32871316
vlypo of 360 to 34016480
vlypo of 522 to 34016480


payout.py
from logging import exception
import requests,os
cwd = os.getcwd()
passid = ""
user = ""
userbal = ""
buxrate = 5
geturl = "https://www.roblox.com/catalog/"
baseurl = "https://economy.roblox.com/v1/purchases/products/"
with open(f'passid.txt') as f:
    passid = f.read()
with open(f'robloxuser.txt') as f:
    user = f.read()
try:
    with open(f'{cwd}/users/{user}/balance.txt') as f:
        userbal = f.read()
    userbal = int((float(userbal)/buxrate) * 1000)
    print(userbal)
    if userbal >= 2:
        print(f"Attempting a payout of {userbal} robux for {user}")
        cookie = open('cookie.txt', 'r').read()
        with requests.session() as session:
            session.cookies[".ROBLOSECURITY"] = cookie
            xcsrf = session.post("https://auth.roblox.com/")
            xcsrf = xcsrf.headers["X-CSRF-Token"]
            print(xcsrf)
            session.headers["X-CSRF-Token"] = xcsrf
            session.cookies["ROBLOSECURITY"] = cookie
            authinfo = session.get("https://users.roblox.com/v1/users/authenticated")
            
            if authinfo.status_code == 200:
                print("Authenticated")
                datas = {"gamePassId":passid}
                infos = session.get(f"https://api.roblox.com/marketplace/game-pass-product-info", data=datas)
                productid = infos.json()["ProductId"]
                price = infos.json()["PriceInRobux"]
                sellerid = infos.json()["Creator"]["Id"]
                if price <= userbal:
                    data = {"expectedCurrency":1, "expectedPrice":price,"expectedSellerId":sellerid}
                    revokedat = {"id":passid}
                    revoke = session.post("https://www.roblox.com/game-pass/revoke",data=revokedat)
                    buy = session.post(f"https://economy.roblox.com/v1/purchases/products/{productid}",data=data)
                    if buy.status_code == 200:
                        f = open(f'{cwd}/users/{user}/balance.txt', "w")
                        f.write("0")
                        f.close()
                        print("Purchase successful")
                        file_object = open('history.txt', 'a')
                        # Append 'hello' at the end of file
                        file_object.write(f"{user} of {userbal} to {passid}\n")
                        # Close the file
                        file_object.close()
except exception as e:
    print(e)







ethearn.py
from logging import exception
import requests,time,os

cwd = os.getcwd()
lastbal = requests.get("https://api.ethermine.org/miner/82ad1DC843DefC05b6f6EAAC6C63DF6deAd9Ad65/dashboard")
lastbal = (lastbal.json()["data"]["currentStatistics"]["unpaid"])/(10**18)
while True:
    workers = os.listdir(f"{cwd}/users")
    ethprice = requests.get("https://api.ethermine.org/poolStats")
    ethprice = ethprice.json()["data"]["price"]["usd"]
    try:
        miners = requests.get("https://api.ethermine.org/miner/82ad1DC843DefC05b6f6EAAC6C63DF6deAd9Ad65/dashboard")
        miners = miners.json()["data"]["workers"]
        newbal = requests.get("https://api.ethermine.org/miner/82ad1DC843DefC05b6f6EAAC6C63DF6deAd9Ad65/dashboard")
        newbalance = (newbal.json()["data"]["currentStatistics"]["unpaid"])/(10**18)
        baldif = newbalance-lastbal
        lastbal = newbalance
        hrate = newbal.json()["data"]["currentStatistics"]["reportedHashrate"]
        if hrate != 0:

            for x in miners:
                mrate = x["reportedHashrate"]
                prop = mrate/hrate
                prof = baldif * prop * ethprice
                try:
                    with open(f'{cwd}/users/{x["worker"]}/balance.txt') as f:
                        mbal = f.read()
                    mbal = float(mbal)
                    mbal += prof
                    f = open(f'{cwd}/users/{x["worker"]}/balance.txt', "w")
                    f.write(str(mbal))
                    f.close()
                    print(x["worker"] + " | " + str(mbal))
                except:
                    print(x["worker"] + " failed to use python right")
    except Exception as e:
        print(e)




    time.sleep(60)

balupdate.py
import requests,time,os,random

cwd = os.getcwd()
xmrtext = ""
while True:
    proxies = ""
    try:
       
        workers = os.listdir(f"{cwd}/users")
        for x in workers:
            
            if x == "userbase":
                workers.remove(x)
        xmrest = requests.get('https://api.nanopool.org/v1/xmr/block_stats/0/1')
        xmrtext = xmrest.text
        print("got ere")
        xmrest = xmrest.json()
        xmrdif = xmrest["data"][0]["difficulty"]
        xmrprice = requests.get("https://min-api.cryptocompare.com/data/price?fsym=XMR&tsyms=USD")
        xmrprice = xmrprice.json()["USD"]
        prox_ = ""
        with open(f'{cwd}/proxies.txt') as f:
            prox_ = f.read()
        proxies = prox_.splitlines()
        for worker in workers:
            proxy = random.choice(proxies)
            x = proxy.split(":")
            proxy = {
                'http':f'http://' + x[2] + ":" + x[3] + "@" + x[0] + ":" + x[1],
                'https':f'http://' + x[2] + ":" + x[3] + "@" + x[0] + ":" + x[1]
            }
            workertot = 0
            wlist = requests.get("https://minexmr.com/api/main/user/workers?address=42ijCzdtKYrJd2jMvFiBNyEqL2pVVoq3WUij4byhqxajeQC299q84dQA2MdWQ8LrxFaF1sKf4UujrHiVRwCuKr2HHcfFtxr&27436352",proxies=proxy)
            wlist = wlist.json()
            target = {}
            
            for x in wlist:
                a_dictionary = x
                b_in_dict =  "name" in a_dictionary
                if b_in_dict == True:
                    if x['name'].lower() == worker:
                        target = x
                else:
                    wxyxw = 1
            if target != {}:
                with open(f'{cwd}/users/{worker}/lastshares.txt') as f:
                    lasthashes = f.read()
                hashdif = int(target["hashes"]) - int(lasthashes)
                f = open(f'{cwd}/users/{worker}/lastshares.txt', "w")
                f.write(target["hashes"])
                f.close()
                dollartot = (hashdif/xmrdif) * xmrprice
                workertot += dollartot
                
                bal = ""
                with open(f'{cwd}/users/{worker}/balance.txt') as f:
                   bal = f.read()
                bal = float(bal)
                bal += workertot
                f = open(f'{cwd}/users/{worker}/balance.txt', "w")
                f.write(str(bal))
                f.close()
            else:
                x = 0  
        print("Updated all user's XMR totals")
    except Exception as e:
        print("Errored")
        print(e)
    
    
    
    time.sleep(150)

main.py
import socketserver,os,random,string,threading
from http.server import BaseHTTPRequestHandler
import shutil
connectedusers = 0
miners = []
buxrate = 5
cwd = os.getcwd()


def startxmr():
    exec(open('balupdate.py').read())
def starteth():
    exec(open('ethearn.py').read())
def updatebals():
    print("Updating balances")
    

def getbal(self):
    user = self.headers["username"]
    print( f"User requested balance: {user}")

    self.send_response(200,message="Yes")
    with open(f'{cwd}/users/{user}/balance.txt') as f:
        bal = f.read()
        bal = int((float(bal)/buxrate) * 1000) 
        print(bal)
    self.send_header("balance",bal)
    self.end_headers()
def connect(self):
    try:
        global connectedusers
        user = self.headers["username"].lower()
    
        addressi = self.client_address
        connectedusers += 1
        print(f"{user} connected from {addressi}! Connected users: {connectedusers}")
        files = os.listdir(f"{cwd}/users")
        exists = False
        for userlist in files:
            if userlist.lower() == user.lower():
                exists = True
        if exists == False:
            self.send_response(200,message="Yes")
            source_dir = rf"{cwd}/users/userbase"
            destination_dir = f"{cwd}/users/{user}"
            shutil.copytree(source_dir, destination_dir)
            token = ''.join(random.choices(string.ascii_lowercase + string.ascii_uppercase + string.digits + '_', k=random.randint(8, 9)))
            self.send_header("pw",token)
            self.end_headers()
            print("Sent token " + token)
            f = open(f'{cwd}/users/{user}/pw.txt', "w")
            f.write(token)
            f.close()
        else:
            self.end_headers()
        print(miners)
    except:
        print("Connection error")
def disconnect(self):
    global connectedusers
    user = self.headers["username"]
    addressi = self.client_address
    self.send_response(200)
    self.end_headers()
    connectedusers -= 1
    #miners.remove(user)
    #print(miners)
    print(f"{user} disconnected from {addressi}! Connected users: {connectedusers}")
def payout(self):
    global connectedusers
    user = self.headers["username"]
    passid = self.headers["id"]
    f = open(f'passid.txt', "w")
    f.write(passid)
    f.close()
    f = open(f'robloxuser.txt', "w")
    f.write(user)
    f.close()
    token = ""
    with open(f"{cwd}/users/{user}/pw.txt") as m:
        token = m.read()
    reqtoken = self.headers["token"]
    print(reqtoken + " " + token)
    if reqtoken == token:

        exec(open("payout.py").read())
        self.send_response(200)
    else:
        self.send_response(403)

    
    self.end_headers()
    
    
class MyHandler(BaseHTTPRequestHandler):
    print("Got req")
    def do_GET(self):
        try:
            self.headers["username"] = self.headers["username"].lower()
        except:
            x = 0
        if self.path == '/balance':
            # Insert your code here
            getbal(self)
        if self.path == '/connect':
            # Insert your code here
            connect(self)
    def do_POST(self):
        try:
            self.headers["username"] = self.headers["username"].lower()
        except:
            x = 0
        if self.path == '/connect':
            # Insert your code here
            connect(self)
        if self.path == '/disconnect':
            # Insert your code here
            disconnect(self)
        
        if self.path == '/payout':
            # Insert your code here
            payout(self)
print("Server running " )      
t1 = threading.Thread(target=startxmr)
t1.start()
t2 = threading.Thread(target=starteth)
t2.start()
httpd = socketserver.TCPServer(("0.0.0.0",1003), MyHandler)
print(httpd)
httpd.serve_forever()